File: website/src/app/services/users.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { PaginatedData } from '../models/paginated-data';
import { Pagination } from '../models/pagination';
import { PostUser } from '../models/post-user';
import { PutUser } from '../models/put-user';
import { User } from '../models/user';

@Injectable({
  providedIn: 'root'
})
export class UsersService {
  constructor(private httpClient: HttpClient) { }

  get(pagination?: Pagination): Observable<PaginatedData<User>> {
    return this.httpClient.get<PaginatedData<User>>('http://localhost:1234/Users?' + 
    'SortColumn=' + (pagination?.sortColumn ?? 'Name') +
    '&Page=' + (pagination?.page ?? 1) +
    '&RowsPerPage=' + (pagination?.rowsPerPage ?? 10) + 
    '&SortAscending=' + (pagination?.sortAscending ?? true));
  }

  getById(userId: number): Observable<User> {
    return this.httpClient.get<User>('http://localhost:1234/Users/' + userId);
  }

  post(dto: PostUser): Observable<User> {
    return this.httpClient.post<User>('http://localhost:1234/Users', dto)
  }

  put(userId: number, dto: PutUser): Observable<User> {
    return this.httpClient.put<User>('http://localhost:1234/Users/' + userId, dto);
  }

  delete(userId: number): Observable<boolean> {
    return this.httpClient.delete<boolean>('http://localhost:1234/Users/' + userId);
  }
}

File: WebAPI/Services/UsersService.cs
using Models;
using Services.DTO;
using Services.Interfaces;
using System.Linq;

namespace Services
{
    public class UsersService : IUsersService
    {
        private readonly Database database;

        public UsersService(Database database)
        {
            this.database = database;
        }

        public bool Delete(int userId)
        {
            User user = this.database.Users
                .Where(u => u.Id == userId)
                .FirstOrDefault();

            if (user == null)
                return false;

            if(user.Login.ToLower() == "admin")
                return false;

            var basketItems = this.database.BasketItems
                .Where(bi => bi.UserId == userId)
                .AsEnumerable();

            this.database.BasketItems.RemoveRange(basketItems);
            this.database.Users.Remove(user);
            this.database.SaveChanges();
            return true;
        }

        public PaginatedData<UserDto> Get(PaginationDto dto)
        {
            PaginatedData<UserDto> paginatedData = new PaginatedData<UserDto>();
            paginatedData.Data = this.database.Users
                .Select(u => new UserDto
                {
                    Id = u.Id,
                    Login = u.Login,
                    Name = u.Name,
                    Surname = u.Surname
                });
            paginatedData.Count = paginatedData.Data.Count();

            switch (dto.SortColumn.ToLower())
            {
                case "login":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(u => u.Login);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(u => u.Login);
                        break;
                    }
                case "name":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(u => u.Name);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(u => u.Name);
                        break;
                    }
                case "surname":
                    {
                        if (dto.SortAscending) paginatedData.Data = paginatedData.Data.OrderBy(u => u.Surname);
                        else paginatedData.Data = paginatedData.Data.OrderByDescending(u => u.Surname);
                        break;
                    }
            }

            paginatedData.Data = paginatedData.Data.Skip((dto.Page - 1) * dto.RowsPerPage).Take(dto.RowsPerPage);
            return paginatedData;
        }

        public UserDto GetById(int userId)
        {
            User user = this.database.Users
                .Where(u => u.Id == userId)
                .FirstOrDefault();

            if (user == null)
                return null;

            return new UserDto
            {
                Id = user.Id,
                Login = user.Login,
                Name = user.Name,
                Surname = user.Surname
            };
        }

        public UserDto Post(PostUserDto dto)
        {
            User user = new User
            {
                Login = dto.Login,
                Password = dto.Password,
                Name = dto.Name,
                Surname = dto.Surname
            };

            if (this.database.Users.Where(u => u.Login.ToLower() == dto.Login.ToLower()).FirstOrDefault() != null)
                return null;

            this.database.Users.Add(user);
            this.database.SaveChanges();

            return new UserDto
            {
                Id = user.Id,
                Login = user.Login,
                Name = user.Name,
                Surname = user.Surname
            };
        }

        public UserDto Put(int userId, PutUserDto dto)
        {
            User user = this.database.Users
                .Where(u => u.Id == userId)
                .FirstOrDefault();

            if (user == null)
                return null;

            user.Name = dto.Name;
            user.Surname = dto.Surname;
            this.database.SaveChanges();

            return new UserDto
            {
                Id = user.Id,
                Login = user.Login,
                Name = user.Name,
                Surname = user.Surname
            };
        }
    }
}
